File: sdk/nodejs/src/testing.ts
/**
 * 无网络依赖的 Mock 客户端，供单元测试使用。
 */

import { AegisGateAPIError, AegisGateSecurityError } from './errors.js';
import type {
  ChatCompletionChunk,
  ChatCompletionRequest,
  ChatCompletionResponse,
  ChatMessage,
  ChatOptions,
  ChatResult,
  HealthResponse,
  ModelListResponse,
  Usage,
} from './types.js';

function lastUserPrompt(messages: ChatMessage[]): string {
  for (let i = messages.length - 1; i >= 0; i--) {
    if (messages[i]!.role === 'user') {
      return messages[i]!.content;
    }
  }
  return '';
}

function fullPromptForUsage(messages: ChatMessage[]): string {
  return messages.map((m) => m.content).join('\n');
}

function estimateUsage(prompt: string, completion: string): Usage {
  const pt = Math.max(1, Math.floor(prompt.length / 4));
  const ct = Math.max(1, Math.floor(completion.length / 4));
  return {
    prompt_tokens: pt,
    completion_tokens: ct,
    total_tokens: pt + ct,
  };
}

let mockIdSeq = 1;
function nextCompletionId(): string {
  return `mock-chatcmpl-${mockIdSeq++}`;
}

/** 单次调用的记录项 */
export interface MockClientCallRecord {
  method: string;
  userMessage?: string;
  request?: ChatCompletionRequest;
  [key: string]: unknown;
}

export class MockAegisGateClient {
  private readonly chatMocks: Array<{ pattern: string; content: string; model: string }> = [];
  private readonly errorMocks: Array<{ pattern: string; aegisCode: string; message: string; statusCode: number }> =
    [];
  private readonly streamMocks: Array<{ pattern: string; chunks: string[]; model: string }> = [];
  private defaultContent: string | null = null;
  private defaultModel = 'mock-model';
  private readonly _calls: MockClientCallRecord[] = [];

  get calls(): MockClientCallRecord[] {
    return [...this._calls];
  }

  get callCount(): number {
    return this._calls.length;
  }

  mockChat(promptPattern: string, options: { content: string; model?: string }): void {
    this.chatMocks.push({
      pattern: promptPattern,
      content: options.content,
      model: options.model ?? 'mock-model',
    });
  }

  mockError(
    promptPattern: string,
    options: { aegisCode: string; message: string; statusCode?: number }
  ): void {
    this.errorMocks.push({
      pattern: promptPattern,
      aegisCode: options.aegisCode,
      message: options.message,
      statusCode: options.statusCode ?? 400,
    });
  }

  mockStream(promptPattern: string, options: { chunks: string[]; model?: string }): void {
    this.streamMocks.push({
      pattern: promptPattern,
      chunks: [...options.chunks],
      model: options.model ?? 'mock-model',
    });
  }

  setDefaultResponse(content: string, model?: string): void {
    this.defaultContent = content;
    this.defaultModel = model ?? 'mock-model';
  }

  reset(): void {
    this.chatMocks.length = 0;
    this.errorMocks.length = 0;
    this.streamMocks.length = 0;
    this.defaultContent = null;
    this.defaultModel = 'mock-model';
    this._calls.length = 0;
  }

  private firstMatch<T extends { pattern: string }>(prompt: string, list: T[]): T | undefined {
    return list.find((e) => e.pattern.length > 0 && prompt.includes(e.pattern));
  }

  private throwIfError(prompt: string): void {
    const hit = this.firstMatch(prompt, this.errorMocks);
    if (!hit) return;
    if (hit.statusCode === 403) {
      throw new AegisGateSecurityError(hit.message, {
        aegisCode: hit.aegisCode,
        errorCode: hit.aegisCode,
        responseBody: JSON.stringify({ code: hit.aegisCode, message: hit.message }),
      });
    }
    throw new AegisGateAPIError(hit.message, {
      statusCode: hit.statusCode,
      errorCode: hit.aegisCode,
      aegisCode: hit.aegisCode,
      responseBody: JSON.stringify({ code: hit.aegisCode, message: hit.message }),
    });
  }

  private makeCompletion(
    content: string,
    model: string,
    messages: ChatMessage[]
  ): ChatCompletionResponse {
    const usage = estimateUsage(fullPromptForUsage(messages), content);
    return {
      id: nextCompletionId(),
      object: 'chat.completion',
      model,
      choices: [
        {
          index: 0,
          message: { role: 'assistant', content },
          finish_reason: 'stop',
        },
      ],
      usage,
    };
  }

  private resolveNonStreamCompletion(request: ChatCompletionRequest): ChatCompletionResponse {
    const messages = request.messages;
    const prompt = lastUserPrompt(messages);
    this.throwIfError(prompt);

    const hit = this.firstMatch(prompt, this.chatMocks);
    if (hit) {
      return this.makeCompletion(hit.content, hit.model, messages);
    }

    if (this.defaultContent !== null) {
      return this.makeCompletion(this.defaultContent, this.defaultModel, messages);
    }

    throw new AegisGateAPIError(
      `未匹配的 prompt，请使用 mockChat / setDefaultResponse: ${JSON.stringify(prompt)}`,
      {
        statusCode: 404,
        errorCode: 'MOCK_UNMATCHED',
        aegisCode: 'MOCK_UNMATCHED',
      }
    );
  }

  async chat(userMessage: string, options: ChatOptions = {}): Promise<ChatResult> {
    this._calls.push({
      method: 'chat',
      userMessage,
      model: options.model,
      temperature: options.temperature,
      maxTokens: options.maxTokens,
      systemPrompt: options.systemPrompt,
    });

    const messages: ChatMessage[] = [];
    if (options.systemPrompt) {
      messages.push({ role: 'system', content: options.systemPrompt });
    }
    messages.push({ role: 'user', content: userMessage });

    const response = this.resolveNonStreamCompletion({
      model: options.model ?? 'gpt-4o',
      messages,
      temperature: options.temperature,
      max_tokens: options.maxTokens,
      stream: false,
    });

    const choice = response.choices[0];
    if (!choice) {
      throw new Error('Chat completion 返回空 choices');
    }

    return {
      content: choice.message.content,
      role: 'assistant',
      finishReason: choice.finish_reason,
      usage: response.usage,
      raw: response,
    };
  }

  async chatCompletions(request: ChatCompletionRequest): Promise<ChatCompletionResponse> {
    this._calls.push({ method: 'chatCompletions', request: { ...request, stream: false } });
    return this.resolveNonStreamCompletion(request);
  }

  async *chatCompletionsStream(
    request: ChatCompletionRequest
  ): AsyncGenerator<ChatCompletionChunk, void, undefined> {
    this._calls.push({ method: 'chatCompletionsStream', request: { ...request, stream: true } });

    const messages = request.messages;
    const prompt = lastUserPrompt(messages);
    this.throwIfError(prompt);

    const sHit = this.firstMatch(prompt, this.streamMocks);
    let chunks: string[];
    let mdl: string;

    if (sHit) {
      chunks = sHit.chunks;
      mdl = sHit.model;
    } else {
      const cHit = this.firstMatch(prompt, this.chatMocks);
      if (cHit) {
        chunks = [cHit.content];
        mdl = cHit.model;
      } else if (this.defaultContent !== null) {
        chunks = [this.defaultContent];
        mdl = this.defaultModel;
      } else {
        throw new AegisGateAPIError(`未匹配的流式 prompt: ${JSON.stringify(prompt)}`, {
          statusCode: 404,
          errorCode: 'MOCK_UNMATCHED',
          aegisCode: 'MOCK_UNMATCHED',
        });
      }
    }

    const id = nextCompletionId();
    const n = chunks.length;
    for (let i = 0; i < n; i++) {
      const text = chunks[i]!;
      yield {
        id,
        object: 'chat.completion.chunk',
        model: mdl,
        choices: [
          {
            index: 0,
            delta: { content: text },
            finish_reason: i === n - 1 ? 'stop' : null,
          },
        ],
      };
    }
  }

  async listModels(): Promise<ModelListResponse> {
    this._calls.push({ method: 'listModels' });
    return {
      object: 'list',
      data: [{ id: 'mock-model', object: 'model', owned_by: 'aegisgate-mock' }],
    };
  }

  async health(): Promise<HealthResponse> {
    this._calls.push({ method: 'health' });
    return { status: 'ok', version: 'mock' };
  }
}

File: sdk/go/testing.go
package aegisgate

import (
	"context"
	"fmt"
	"strings"
	"sync"
)

// MockCall 单次 Mock 客户端调用的记录
type MockCall struct {
	Method string
	Prompt string
	Model  string
}

type mockChatRule struct {
	pattern string
	content string
}

type mockErrRule struct {
	pattern   string
	aegisCode string
	message   string
}

// MockClient 无 HTTP 的测试替身，与 Client 的关键方法签名对齐
type MockClient struct {
	mu sync.Mutex

	chatRules []mockChatRule
	errRules  []mockErrRule

	defaultContent string
	defaultOK      bool

	nextID int
	calls  []MockCall
}

// NewMockClient 创建 Mock 客户端；opts 与 NewClient 相同，用于保持配置 API 一致（当前不用于网络）
func NewMockClient(opts ...Option) *MockClient {
	_ = NewClient(opts...)
	return &MockClient{}
}

// MockChat 按子串匹配注册非流式回复
func (m *MockClient) MockChat(promptPattern, responseContent string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.chatRules = append(m.chatRules, mockChatRule{pattern: promptPattern, content: responseContent})
}

// MockError 按子串匹配注册错误（返回 *APIError，含 AegisCode）
func (m *MockClient) MockError(promptPattern string, aegisCode string, message string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.errRules = append(m.errRules, mockErrRule{pattern: promptPattern, aegisCode: aegisCode, message: message})
}

// SetDefaultResponse 未匹配规则时的默认回复内容
func (m *MockClient) SetDefaultResponse(content string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.defaultContent = content
	m.defaultOK = true
}

func mockLastUserPrompt(msgs []Message) string {
	for i := len(msgs) - 1; i >= 0; i-- {
		if msgs[i].Role == "user" {
			return msgs[i].Content
		}
	}
	return ""
}

func mockFullPrompt(msgs []Message) string {
	var b string
	for i, m := range msgs {
		if i > 0 {
			b += "\n"
		}
		b += m.Content
	}
	return b
}

func mockEstimateUsage(prompt, completion string) *Usage {
	pt := len(prompt) / 4
	if pt < 1 {
		pt = 1
	}
	ct := len(completion) / 4
	if ct < 1 {
		ct = 1
	}
	return &Usage{
		PromptTokens:     pt,
		CompletionTokens: ct,
		TotalTokens:      pt + ct,
	}
}

func (m *MockClient) recordLocked(method, prompt, model string) {
	m.calls = append(m.calls, MockCall{Method: method, Prompt: prompt, Model: model})
}

func (m *MockClient) nextChatIDLocked() string {
	m.nextID++
	return fmt.Sprintf("mock-chatcmpl-%d", m.nextID)
}

func firstMockChatMatch(prompt string, rules []mockChatRule) *mockChatRule {
	for i := range rules {
		if rules[i].pattern != "" && strings.Contains(prompt, rules[i].pattern) {
			return &rules[i]
		}
	}
	return nil
}

func firstMockErrMatch(prompt string, rules []mockErrRule) *mockErrRule {
	for i := range rules {
		if rules[i].pattern != "" && strings.Contains(prompt, rules[i].pattern) {
			return &rules[i]
		}
	}
	return nil
}

// ChatCompletions 返回注册的补全结果（非流式）
func (m *MockClient) ChatCompletions(ctx context.Context, req *ChatCompletionsRequest) (*ChatCompletionsResponse, error) {
	if req.Stream {
		return nil, &AegisGateError{Message: "use ChatCompletionsStream for streaming"}
	}
	_ = ctx

	m.mu.Lock()
	defer m.mu.Unlock()

	prompt := mockLastUserPrompt(req.Messages)
	m.recordLocked("ChatCompletions", prompt, req.Model)

	if hit := firstMockErrMatch(prompt, m.errRules); hit != nil {
		return nil, &APIError{
			Message:    hit.message,
			StatusCode: 400,
			ErrorCode:  hit.aegisCode,
			AegisCode:  hit.aegisCode,
		}
	}

	if hit := firstMockChatMatch(prompt, m.chatRules); hit != nil {
		content := hit.content
		return &ChatCompletionsResponse{
			ID:     m.nextChatIDLocked(),
			Object: "chat.completion",
			Model:  req.Model,
			Choices: []Choice{
				{
					Index: 0,
					Message: Message{
						Role:    "assistant",
						Content: content,
					},
					FinishReason: "stop",
				},
			},
			Usage: mockEstimateUsage(mockFullPrompt(req.Messages), content),
		}, nil
	}

	if m.defaultOK {
		return &ChatCompletionsResponse{
			ID:     m.nextChatIDLocked(),
			Object: "chat.completion",
			Model:  req.Model,
			Choices: []Choice{
				{
					Index: 0,
					Message: Message{
						Role:    "assistant",
						Content: m.defaultContent,
					},
					FinishReason: "stop",
				},
			},
			Usage: mockEstimateUsage(mockFullPrompt(req.Messages), m.defaultContent),
		}, nil
	}

	return nil, &APIError{
		Message:    fmt.Sprintf("未匹配的 prompt，请使用 MockChat / SetDefaultResponse: %q", prompt),
		StatusCode: 404,
		ErrorCode:  "MOCK_UNMATCHED",
		AegisCode:  "MOCK_UNMATCHED",
	}
}

// Health 返回固定健康数据
func (m *MockClient) Health(ctx context.Context) (*HealthResponse, error) {
	_ = ctx
	m.mu.Lock()
	defer m.mu.Unlock()
	m.recordLocked("Health", "", "")
	return &HealthResponse{Status: "ok", Version: "mock"}, nil
}

// ListModels 返回固定模型列表
func (m *MockClient) ListModels(ctx context.Context) (*ModelListResponse, error) {
	_ = ctx
	m.mu.Lock()
	defer m.mu.Unlock()
	m.recordLocked("ListModels", "", "")
	return &ModelListResponse{
		Object: "list",
		Data: []ModelInfo{
			{ID: "mock-model", Object: "model", OwnedBy: "aegisgate-mock"},
		},
	}, nil
}

// Calls 返回调用记录的副本
func (m *MockClient) Calls() []MockCall {
	m.mu.Lock()
	defer m.mu.Unlock()
	out := make([]MockCall, len(m.calls))
	copy(out, m.calls)
	return out
}

// CallCount 返回调用次数
func (m *MockClient) CallCount() int {
	m.mu.Lock()
	defer m.mu.Unlock()
	return len(m.calls)
}

// Reset 清空规则与调用记录
func (m *MockClient) Reset() {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.chatRules = nil
	m.errRules = nil
	m.defaultOK = false
	m.defaultContent = ""
	m.nextID = 0
	m.calls = nil
}
